Skip receipt page if placing free order

xitij2000 opened the pull request:
This PR changes the behaviour of ecommerce to skip the receipt page when a coupon redemption doesn't require any payment. When placing a free order the receipt page will be skipped and users will be directed to the LMS dashboard page on redemption instead. 

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:
The regular test suite can be used for testing this change by running the tests for  ``ecommerce.extensions.checkout``

**Reviewers**
- [ ] @kaizoku
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [OSPR-2295](https://openedx.atlassian.net/browse/OSPR-2295) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/ecommerce/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented:
@edx/learner Is this the right way to do this?

nedbat commented:
@xitij2000 Ping me when this is ready for review.

Ayub-Khan commented:
a better way to do that would be using waffle flag. As in ecommerce all the such things are done using waffle flags.

xitij2000 commented:
@Ayub-Khan Thanks for pointing that out! I have updated the code to use a waffle flag.

nedbat commented:
@xitij2000 is this still a work-in-progress, or is it ready for review?

xitij2000 commented:
@nedbat  Sorry for the delay! I will try to get this finished today and get it reviewed internally and ping you as soon as possible.

kaizoku commented:
:+1:

- [✓] I tested this: (deployed in devstack, enabled waffle switch, and checked out 100% discount coupon)
- [✓] I read through the code

xitij2000 commented:
@nedbat Sorry for the delay! This is ready for review now.

mduboseedx commented:
@edx/learner Could this please be reviewed?

mikix commented:
I know we told you that waffle flags are decent, but actually, we are trying to reduce our dependence on them (they are not as easy to document and make our code more spaghetti like than we want).

Instead of waffle flags, we are hoping to eventually move optional features to plugins (that can be installed or not) or maybe config values if we do have to have tweakability.

However in this case, my engineering brain thinks maybe this is just a change we always want? Though I admit I don't have much context here. I'll kick this to the Product team to see if they see value in the old way (receipt for free orders). <- @edx/learner-product

xitij2000 commented:
@mikix Thaks for the additional context! 
If it makes more sense to skip the receipt for free orders then perhapes the condition of this flag can be reversed. So we can have a ``SHOW_RECEIPT_FOR_FREE_ORDERS`` flag/setting. 
Either way seems fine to me. I look forward to hearing from the Product team.

mikix commented:
@xitij2000 to be clear, I was proposing dropping the flag altogether and just hard-coding this behavior.

marcotuts commented:
One question here I have is what is the current behavior with & without the flag for showing the payment in the order history tab of the account settings area. 

Thanks for flagging the product questions here. I will defer to Deen on this one, I can connect with him offline and update this flag with a question. Independent of his feedback, an answer to the question above would be helpful context. Thanks!

xitij2000 commented:
@marcotuts  Sorry for the incredibly late reply, but I think after this change you will still be able to see details in order history.

marcotuts commented:
No flag and defaulting to skipping receipt for free orders makes sense to me. This can move to engineering review next.

xitij2000 commented:
@mikix Thanks! I have rebased the code on the latest master, and the tests are passing now. I also tested it manually to see if it still works after the rebase.

mikix commented:
@robrap can someone from the ecommerce guild look at this & push if it's good?

robrap commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -90,11 +89,7 @@ def get_redirect_url(self, *args, **kwargs):
-                    site_configuration=order.site.siteconfiguration
-                )
-                url = site.siteconfiguration.build_lms_url(receipt_path)
+                url = get_lms_dashboard_url()
```
The docstring for this class needs to be updated appropriately.  

Also, is it safe to assume that this method is only called once it is determined the basket was free?  If so, maybe add the following to this method:
```
Returns the redirect url for the free basket.
```

xitij2000 replied:
@robrap Earlier in this method it checks if the basket total is 0, and raises an error if it isn't, so by the time we reach this code I think it's safe to assume it's a free order. 

I've updated the docstring for the class, however since the method is called {{get_redirect_url}} so not sure adding that docstring will really help!

robrap commented:
@xitij2000: You can squash and rebase.  However, I don't plan on merging until Thursday morning after the holiday here, so you may want to wait and rebase then.

Also, have you ensured that there are no side-effects on the receipt page other than displaying the receipt?  Tracking log events or anything else?  Just want to ensure that there are no unintended consequences.  Thanks.

xitij2000 commented:
@robrap  I had a closer look at the replaced code, and it's just code for building URLs, nothing that would create any kind of log output of any kind. 

I've also rebased and squashed the commits.

edx-webhook commented:
@xitij2000 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/ecommerce&entry.867055334=https://github.com/edx/ecommerce/pull/1676&entry.1484655318=https://github.com/xitij2000&entry.752974735=2018-03-14+12:34&entry.1917517419=2018-07-06+13:02&entry.2133058324=Yes) so we can improve your experience in the future.

xitij2000 commented:
@robrap  Thanks!

robrap commented:
@xitij2000: Unfortunately, I'm not sure if I will need to revert this or not.  We have some e2e tests that relied on the receipt page being there that are now failing, and not sure when someone will be able to pick up the work to fix them and/or if you would be able to.  More next week...

xitij2000 commented:
@robrap Sure will try to help in whatever way I can.

robrap commented:
Looks like I will need to revert on Monday. Beyond the end-to-end tests failing, I think the user is getting to the dashboard with no message and no course, presumably beating out fulfillment.

xitij2000 commented:
@robrap OK. Would it be more acceptable if I put this behaviour behind a waffle flag/switch or setting like it was originally?

robrap commented:
My personal opinion is that this requires some thought, but it’s not my call. As written, the user sometimes lands on the dashboard without the newly enrolled course being available and with no message. The user has to refresh the page for it to appear. This isn’t a great user experience. It seems that a consequence of the receipt page is adding enough delay to generally avoid this problem.

A more complete solution would be to show a success message that could tell the user what they need to do, or automates the appropriate reload, but as you could imagine this could get complicated, especially because it is across services. A related alternative is to embed this into the receipt page with a timer countdown that automatically redirects, and includes a link to manually go to dashboard before the timer. Probably there are other ideas as well. I just don’t think the current issue is good for users as-is, whether it is behind a flag or not.

xitij2000 commented:
I see. That does sound like a more complex issue to solve! Thanks for all your help. I'll see if there is any way to take this forward in a way that works, or if it's even worth it.

robrap commented:
@xitij2000: I don't plan on putting any more time into this as-is.  If you want to consider approaches and discuss alternatives, let me know.  Thanks.

xitij2000 commented:
@robrap I checked up on this a bit, and the main requirement of our client is that for users enrolling via an enrollment code should not be confronted with a price point + 100% discount. So simply hiding those details in case of a free enrollment would be sufficient. 

Would it be acceptable to add the ability to hide the pricepoint in case of free orders behind a flag?

robrap commented:
@marcotuts: Can you get the proper Product review for the above comment?  Again, it would be great to avoid a flag.

xitij2000 commented:
I figured a flag would be best because I can understand a genuine need for showing the price in the receipt even for a 100% discount.

xitij2000 commented:
@marcotuts Does the approach in my previous comment seem reasonable? 

Essentially, if a user is joining a course using a code that reduces the price to 0, then don't show them the price at all. This could behind a flag, or (as @robrap mentioned above) could be the new default behaviour. 

Would love to know if this approach is acceptable/feasible so we can start work on it.
